trackpy/find_link.py

```diff
@@ -424,6 +422,117 @@ def __init__(self, t, pos, id=None, extra_data=None):
         self.relocate_neighbors = []
 
 
+def _default_coord_mapping(search_range, level):
+    """ Convert a list of Points to an ndarray of coordinates """
+    return _points_to_arr(level) / search_range
+
+
+def _wrap_predictor(predictor, t):
+    def coord_mapping(search_range, level):
+        return list(predictor(t, level)) / search_range
+    return coord_mapping
+
+
+class TreeFinder(object):
+    def __init__(self, points, search_range):
+        """Takes a list of particles."""
+        self.ndim = len(search_range)
+        self.search_range = np.atleast_2d(search_range)
+        if not isinstance(points, list):
+            points = list(points)
+        self.points = points
+        self.set_predictor(None)
+        self.rebuild()
+
+    def __len__(self):
+        return len(self.points)
+
+    def add_point(self, pt):
+        self.points.append(pt)
+        self._clean = False
+
+    def set_predictor(self, predictor, t=None):
+        """Sets a predictor to the TreeFinder
+
+        predictor : function, optional
+
+            Called with t and a list of N Point instances, returns their
+            "effective" locations, as an N x d array (or any iterable).
+            Used for prediction (see "predict" module).
+        """
+        if predictor is None:
+            self.coord_mapping = _default_coord_mapping
+        else:
+            self.coord_mapping = _wrap_predictor(predictor, t)
+        self._clean = False
+
+    @property
+    def kdtree(self):
+        if not self._clean:
+            self.rebuild()
+        return self._kdtree
+
+    def rebuild(self):
+        """Rebuilds tree from ``points`` attribute.
+
+        coord_map : function, optional
+
+            Called with a list of N Point instances, returns their
+            "effective" locations, as an N x d array (or list of tuples).
+            Used for prediction (see "predict" module).
+
+        rebuild() needs to be called after ``add_point()`` and
+        before tree is used for spatial queries again (i.e. when
+        memory is turned on).
+        """
+        self._clean = False
+        if len(self.points) == 0:
+            self._kdtree = None
+        else:
+            self._kdtree = cKDTree(self.coords_mapped, 15)
+        # This could be tuned
+        self._clean = True
+
+    @property
+    def coords(self):
+        return _points_to_arr(self.points)
+
+    @property
+    def coords_mapped(self):
```

Comment:
It looks like `coords_mapped` should be used only once per frame. Is that true? Depending on the predictor this can be an expensive operation.